cmd/fpga_plugin/README.md

```diff
@@ -1,120 +1,290 @@
+        * [Run FPGA device plugin in af mode](#run-fpga-device-plugin-in-af-mode)
+        * [Run FPGA device plugin in region mode](#run-fpga-device-plugin-in-region-mode)
+* [Next steps](#next-steps)
+* [Orchestrated mode](#orchestrated-mode)
 
--   [FPGA device plugin](README.md) (this module)
+# Introduction
+
+This FPGA device plugin is part of a collection of Kubernetes components found within this
+repository that enable integration of Intel FPGA hardware into Kubernetes.
+
+The following hardware platforms are supported:
+
+- Intel Arria 10
+- Intel Stratix 10
+
+The components support the [Open Programmable Acceleration Engine (OPAE)](https://opae.github.io/latest/index.html)
+and OpenCL application interfaces.
+
+The components together implement the following features:
+
+- discovery of pre-programmed accelerator functions
+- discovery of programmable regions
+- orchestration of FPGA programming
+- access control for FPGA hardware
+
+# Component overview
+
+The following components are part of this repository, and work together to support Intel FPGAs under
+Kubernetes:
+
+-   [FPGA device plugin](README.md) (this component)
+
+    A Kubernetes [device plugin](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/)
+    that discovers available FPGA resources on a node and advertises them to the Kubernetes control plane
+    via the node kubelet.
+
+-   [FPGA admission controller webhook](../fpga_admissionwebhook/README.md)
+
+    A Kubernetes [admission controller webhook](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/)
+    which can be used to dynamically convert logical resource names in pod specifications into actual FPGA
+    resource names, as advertised by the device plugin.
+
+    The webhook can also set environment variables to instruct the CRI-O prestart hook to program the FPGA
+    before launching the container.
+
+-   [FPGA CRI-O prestart hook](../fpga_crihook/README.md)
+
+    A [CRI-O](https://github.com/cri-o/cri-o) prestart hook that, upon instruction from the FPGA admission
+    controller, allocates and programs the FPGA before the container is launched.
+
+The repository also contains an [FPGA helper tool](../fpga_tool/README.md) that may be useful during
+development, initial deployment and debugging.
+
+# FPGA modes
+
+The FPGA plugin set can run in one of two modes:
+
+- `af`/`preprogrammed` mode, where the FPGA bitstreams are already loaded
+  onto the FPGA, and the plugins discover and advertises the existing
+  Accelerator Functions (AF).
+- `region`/`orchestration` mode, where the plugins locate and advertise
+  regions of the FPGA, and facilitate programing of those regions with the
+  requested bistreams.
+
+The example YAML deployments described in this document only currently support
+`af`/`preprogrammed` mode. To utilise `region`/`orchestrated` mode, either modify
+the existing YAML appropriately, or deploy 'by hand'.
+
```

Comment:
I'm not sure I fully agree with this.

[This yaml](https://github.com/grahamwhaley/intel-device-plugins-for-kubernetes/blob/20200117_fpga_readme/deployments/fpga_plugin/fpga_plugin.yaml) uses initcontainer which makes little sense to use in preprogrammed mode.

Plugin runs in af/preprogrammed mode by default, but  it can be changed by [setting node annotation](https://github.com/grahamwhaley/intel-device-plugins-for-kubernetes/blob/20200117_fpga_readme/demo/screencast-fpga-orchestrated-dfl.sh#L111).

As preprogrammed mode is not fully supported by the webhook, I'd suggest to concentrate on describing orchestrated mode here.

Reply:
The root issue aiui is that we only supply one daemonset example (which in my mind is far the preferable way to get somebody up and running and show an example deployment), but that daemonset, as you say, includes the initcontainer but runs in `af` mode by default - which is a 'conflict of interests'.

If we need to focus on orchestrated mode, then I think we need another daemonset with the correct settings for that mode (enable region mode, remove the initcontainer). wdyt?

Ideally, I'd not be writing that daemonset, as I don't have the experience with the FPGA setup, testing, and nor do I have easy access to the hardware ;-)

Sooo, I was trying to do the best with the documentation with the example files at hand...

Reply:
How about describing deployment in both modes? The daemonset should not be changed, it can be used in both modes. In orchestration mode CRI-O must be used and orchestrated mode should be set as an annotation.

Reply:
OK, have added two deployment sub-sections, one for each mode.